### PR TITLE
Integrate PCRE capture handling

### PR DESCRIPTION
**Pull request by silentbicycle**
This integrates a new approach for handling captures in one _or more_ PCRE regexes.

This a draft because:

- A few tests in `capture_test_case_list.c` are currently marked as `SHOULD_SKIP`, such as the one for `^(?:($|x))+$`. They are all variants of the same underlying issue, which is that repetition with a `+` surrounding multiple ALT cases where earlier subtrees can match without consuming input is currently handled incorrectly -- I have confirmed that my intended fix for the generated capture resolution opcodes works, but I'm still working on flagging that case without false positives.
- Similarly, code generation is not implemented for `{count,}` where count is > 1. This isn't a deep issue, `(?:abc){3,}` should desugar to `(?:abc){3}(?:abc)*`, I just haven't set it up yet.
- Some of the modes for `re` aren't enabled, such as capture resolution for multiple files.
- The capture VM's path node sharing needs more testing, particularly in combination with the collapsed zero prefix.
- We should be able to generate worst case memory usage info at regex compile time, which could be used by the caller to stack-allocate / reuse a preallocated buffer, but I haven't set that up yet.
- Furthermore, we should be able to do capture resolution in a *fixed* preallocated amount of memory, if replay of the "correct" match prefix is interleaved with search (once all surviving threads agree on the prefix), with incremental resetting when resuming from the all-zero prefix. I haven't implemented this yet either, because it's likely to be fairly complicated, but it may be worth doing later once we have more information about typical memory usage in the fleet.

# Test Data

Getting test data from `capture_test_case_list.c`: Build and run the tests, then run `build/tests/capture/run_test_case_list -vvvvvvvvv` (yes, nine '-v's). This will write out the generated programs and character classes for each test into `prog_output`, in a format that should be relatively easy to scrape. Sorry about the hackiness here -- I'm going to replace libfsm's output generation next, which will give a much better interface for converting regexes to standalone test data.

The test case output looks like this:
```
==== test_case 38
regex "$(x?)(y?)(z?)", input "a", match SHOULD_MATCH, no_nl 0, count 4: 0:[1, 1] 1:[1, 1] 2:[1, 1] 3:[1, 1]
fsm_capture_dump_active_for_ends:
 -- state 0: value 0
 -- state 0: value 1
 -- state 0: value 2
 -- state 0: value 3

==== fsm_capture_dump_programs:
# program 0, capture_count 4, base 0
0: split cont 3 new 1
1: charclass 0 -> [ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff]
2: jmp 0
3: save 0 (cap 0, start)
4: anchor end
5: save 2 (cap 1, start)
6: split cont 7 new 8
7: char 0x78 (x)
8: save 3 (cap 1, end)
9: save 4 (cap 2, start)
10: split cont 11 new 12
11: char 0x79 (y)
12: save 5 (cap 2, end)
13: save 6 (cap 3, start)
14: split cont 15 new 16
15: char 0x7a (z)
16: save 7 (cap 3, end)
17: save 1 (cap 0, end)
18: match
char_class 0: 0xffffffffffffffff 0xffffffffffffffff 0xffffffffffffffff 0xffffffffffffffff

fsm_capture_dump_program_end_mapping:
 -- state 0: value 0
```
The `fsm_capture_dump_program_end_mapping` and `fsm_capture_dump_active_for_ends` can be ignored, if testing a port of the capture VM to another language you only need the following:

- `regex "$(x?)(y?)(z?)", input "a", match SHOULD_MATCH, no_nl 0, count 4: 0:[1, 1] 1:[1, 1] 2:[1, 1] 3:[1, 1]`
- the program (opcodes are described in `capture_vm_program.h`)
- its char_class list (there may be zero)

and then check that executing the program with the same input sets the capture buffers correctly, here to:
```
- 0: 1,1
- 1: 1,1
- 2: 1,1
- 3: 1,1
```
The capture VM is in `src/libfsm/capture_vm*`.

### CODE REVIEW

**Comment by silentbicycle** [comment]
There are two CI failures, both related to the `{3,}` case I mentioned above returning a "not yet implemented" error. That will get fixed before I change this from a draft PR.

File: include/re/re.h [hunk 1]
```diff
@@ -64,6 +65,8 @@ enum re_errno {
 	RE_EERRNO       =  1 | RE_MISC,
 	RE_EBADDIALECT  =  2 | RE_MISC,
 	RE_EBADGROUP    =  3 | RE_MISC,
+	RE_EUNSUPCAPTUR =  4 | RE_MISC,
```
Comment (katef) [inline include/re/re.h:68]: hello Ken

**Comment by silentbicycle** [comment]
The new commits I just pushed address the capture tests that were marked as `SKIP`. All tests are passing locally, but this needs further fuzzer burn-in time.

**Comment by silentbicycle** [comment]
This is going to go in PRs for the public upstream repo (starting with https://github.com/katef/libfsm/pull/440), then will merge here as part of a sync from upstream.
